sdk/nodejs/resource.ts

```diff
@@ -486,6 +508,33 @@ export interface CustomTimeouts {
     delete?: string;
 }
 
+/**
+ * ResourceTransformation is the callback signature for the `transformations` resource option.  A
+ * transformation is passed the same set of inputs provided to the `Resource` constructor, and can
+ * optionally return back alternate values for the `props` and/or `opts` prior to the resource
+ * actually being created.  The effect will be as though those props and opts were passed in place
+ * of the original call to the `Resource` constructor.  If the transformation returns undefined,
+ * this indicates that the resource will not be transformed.
+ */
+export type ResourceTransformation =
+    (type: string, name: string, props: Inputs, opts: ResourceOptions) => ResourceTransformationResult | undefined;
```

Comment:
So if I want to modify a property, presumably I need to do it like
```
pulumi.output(props).apply(...)
```
?

Reply:
Yes - or `pulumi.output(props["myprop"]).apply(myprop => ...)` in the case that you just want to overwrite a single property in a way that is conditional on the value passed in.

This is slightly more cumbersome to author - but is more expressive - as it allows things like injecting new dependencies.  For a low level power-user feature like this - my feeling was that  providing flexibility/expressivity is more valuable than providing ease-of-authoring.  

I've added a new test that is motivated by the scenario in https://github.com/pulumi/pulumi/issues/2068#issuecomment-533839439 which shows how this can be taken advantage of for a real (and fairly complex) scenario involving injecting a dependency on one child into another child. 

Also - FWIW - in most common cases I've come up with - the transformation is overwriting the value or injecting a new default - so doesn't actually need to look at the existing input value.